FLAIR-HUB dataset

gatienc opened the pull request:
The [FLAIR-HUB](https://github.com/IGNF/FLAIR-HUB) dataset is a large-scale, multimodal dataset for land cover and crop mapping from the French National Institute of Geographical and Forest Information (IGN). It builds upon and includes the FLAIR#1 and FLAIR#2 datasets, expanding them into a unified resource with very-high-resolution annotations.

This PR adds the [FLAIRHUB](https://ignf.github.io/FLAIR/FLAIR-HUB/flairhub.html) mono-temporal part of the dataset.

<img width="1589" height="365" alt="image" src="https://github.com/user-attachments/assets/4f966f76-6227-4ee4-a115-9c044ec8b838" />

Let me know if anything has to changed. the monotemporal part of the dataset is already very solid and can be used independently. I plan to work on the time-series part of the dataset once it will be implemented in Torchgeo.

## Note 
As the dataset size is huge ( total of ~750 GB) I wasn't able to test training on the full dataset but it is possible to work on a subset of the dataset by commenting most of the domain years dict in  Torchgeo/datasets/flairhub.py#L174)

## Acknowledgments
Special thanks to the work done in #2394, which served as a starting point for this implementation.

## Todo
- [x] Implement the time-series part of the dataset
- [x] Add the historical part of the dataset

adamjstewart commented:
> The LPIS (crop) dataset is labeled with three hierarchical levels. Currently, the paper only utilizes the first level for model training, ignoring the second and third levels which offer deeper classification. What would be the best method to implement a hierarchical classification system ? Is there any standard datasets I could use as an example?

EuroCrops and maybe TreeSatAI also have hierarchical classification, although this isn't exposed via TorchGeo. Simplest way to implement would be a Dataset parameter letting you select the hierarchy level. This isn't really full hierarchical classification though. Does torchmetrics have any way to compute metrics for hierarchical predictions?

gatienc commented:
I removed duplicate TIF files from history by squashing them.

The temporal data is not used in the datamodule as I haven't implemented the paper's model.

@adamjstewart The dataset should now be ready for review

adamjstewart commented on torchgeo/datasets/flairhub.py:
```diff
@@ -0,0 +1,1361 @@
+    sentinel1_desc_ts_bands: tuple[str, str] = ('VV', 'VH')
+
+    # HuggingFace dataset URL
+    url_base: ClassVar[str] = (
```
ClassVar isn't needed because strings are immutable

adamjstewart commented on torchgeo/datasets/flairhub.py:
```diff
@@ -0,0 +1,1361 @@
+
+    # HuggingFace dataset URL
+    url_base: ClassVar[str] = (
+        'https://huggingface.co/datasets/IGNF/FLAIR-HUB/resolve/main/data'
```
```suggestion
        'https://hf.co/datasets/IGNF/FLAIR-HUB/resolve/e8ed7981d488508aa70bb05c37cf6585432b7d5f/data'
```
Git commit hash is required for stable downloads

adamjstewart commented on torchgeo/datasets/flairhub.py:
```diff
@@ -0,0 +1,1361 @@
+
+    def __init__(
+        self,
+        root: Path = Path('data'),
```
Let's default to a str to match all other existing datasets.

adamjstewart commented on torchgeo/datasets/flairhub.py:
```diff
@@ -0,0 +1,1361 @@
+
+        Returns:
+            dictionary containing tensors for each modality
+            keys are the modality names : 'mask', 'aerial_rgbi', 'spot_rgbi', 'dem_elev', 'aerial_rlt_pan', 'sentinel2_ts', 'sentinel2_msk_sc', 'sentinel1_asc_ts', 'sentinel1_desc_ts'
```
Newline is removed, so without a period on the first sentence this won't make sense.

adamjstewart commented on torchgeo/datasets/flairhub.py:
```diff
@@ -0,0 +1,1361 @@
+
+    download_link: str = 'https://storage.gra.cloud.ovh.net/v1/AUTH_366279ce616242ebb14161b7991a8461/defi-ia/flair_hub/FLAIR-HUB_TOY_DATASET.zip'
+
+    def __init__(
```
Is this `__init__` different from the super class? Wondering if we can remove this.

adamjstewart commented on torchgeo/datamodules/flairhub.py:
```diff
@@ -0,0 +1,287 @@
+        num_workers: int = 0,
+        n_test_domains: int = 16,
+        train_val_split: float = 0.8,
+        random_state: int = 0,
```
None of our other datamodules have this. Maybe we should just hard-code it to 0?

adamjstewart commented on tests/datamodules/test_flairhub.py:
We are trying to remove this directory, as datamodule tests aren't actually useful. Instead, we should test all datamodules with their corresponding trainers. I'll add a README to make this more clear.

adamjstewart replied:
If we remove the handling for wrong input, we can get rid of this file. I try to use type hints to at least statically check that all inputs are valid.

gatienc replied:
How should we handle the tests for the official splits of the dataset then? (If we keep it in datamodules, otherwise : https://github.com/torchgeo/torchgeo/pull/3181#discussion_r2752004254)

adamjstewart reviewed (changes requested):
Currently ~150 files and ~2.7k lines of code. If we can get this to ~10 files and ~500 lines of code, I would feel much more comfortable about merging this. Whether or not the dataset works is one thing, but whether or not it is maintainable is another.

There is a lot of complexity added by the dataset being multimodal. If we return a single 'image' with all bands stacked together, that solves most of the complexity problems. We also can't have any tests that download anything during CI, so make sure we aren't downloading test split files or things like that.

Also needs documentation, updates to `docs/api/datasets.rst`, `docs/api/datamodules.rst`, and `docs/api/datasets/non_geo_datasets.csv`.

gatienc commented on torchgeo/datasets/flairhub.py:
Flair-Hub is a direct improvement over Flair#1 and Flair#2, already incorporating all their data with fixes and additional content. I don't know if there is a need to support Flair#1 and Flair#2 then.

gatienc committed:
feat: Add FLAIRHUB dataset support

gatienc committed:
adding docs for flair-hub

gatienc committed:
test: adding coverage for plotting of lpis 2 and 3 mask and adding geometry to mocked GDF to avoid Attribute error in minimal tests

gatienc committed:
removing unecessary comments and info in docstrings

gatienc committed:
update of data.py

- renamed data/flairhub to data/flair
- adding SPOT_RGBI to the modalities generated for all domain-years to trigger resize operation

gatienc committed:
adding SPOT_RGBI to the test config to trigger resize operation

gatienc committed:
fix: SENTINEL2_MSK_SC test data

SENTINEL2_MSK_SC was wrongly generated, fixing its generation code and regenerating data.py

gatienc committed:
fix: training with SPOT, forgot update in_channels

gatienc committed:
test: simplifying test

- removing any redundant test
- avoiding code duplication when possible
- removed useless check

gatienc commented:
I’ve implemented all the requested changes, except for the band naming issue.

> "Most other datasets use bands [B1, B2, B3, ...] or [VV, VH] or even [VV, VH, B1, B2, B3, ...]. This gives maximum flexibility for cases where you want to test a pre-trained model but it has its bands in a weird order or only a subset of bands are supported.
> 
> Only a few datasets have sensors S1, S2, etc. I would like to slowly phase these out and replace them with the above."


For this dataset, since the same band numbers refer to different spectral ranges across sensors, I’m unsure how to uniquely name the bands.
One option could be: bands = [SPOT_B1, SPOT_B2, ... , S1_ASC_VV,...]
But this option doesn't show which dataset are monotemporal and which are multi-temporal so I assume this is better to avoid confusion
bands = [SPOT_B1, SPOT_B2, ... , TS_S1_ASC_VV, ...]
However, this approach feels overly verbose. Any suggestions or preferences for how to handle this?

gatienc commented:
Also I suppose the version added should be 0.10.0 now?

gatienc committed:
updating version added

gatienc committed:
Fixing type hints for ty

gatienc committed:
ruf formatting

adamjstewart commented on tests/datasets/test_flair.py:
```diff
@@ -0,0 +1,311 @@
+            )
+
+    def test_len(self, dataset: FLAIRHUB | FLAIRHUBToy) -> None:
+        assert len(dataset) > 0
```
I would rather assert a specific length, we can play around with the fake test data and test params to ensure that this number is always the same.

adamjstewart commented on tests/datasets/test_flair.py:
```diff
@@ -0,0 +1,311 @@
+            ('crop_type_2', ['AERIAL_RGBI'], 'LPIS 2 (crop_type_2)'),
+            ('crop_type_3', ['AERIAL_RGBI'], 'LPIS 3 (crop_type_3)'),
+        ],
+    )
```
We could inherit the dataset fixture instead of parametrizing this separately.

adamjstewart commented on tests/datasets/test_flair.py:
```diff
@@ -0,0 +1,311 @@
+        assert fig is not None
+        plt.close()
+
+    def test_getitem_with_transforms(self, tmp_path: Path) -> None:
```
We don't test this for any other datasets, could remove if it's not needed for coverage

adamjstewart commented on tests/trainers/test_segmentation.py:
```diff
@@ -94,7 +98,6 @@ def test_trainer(
         match name:
             case 'ftw':
                 pytest.importorskip('pyarrow')
-
```
Can you undo this change?

adamjstewart commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1384 @@
+                if domain_year not in allowed_domain_years:
+                    continue
+
+            file_dict: dict[str, Any] = {}
```
Trying to avoid `Any` because it disables the type checker, could we use a TypedDict here?

adamjstewart reviewed (changes requested):
Apologies for the delayed review, I think I'm finally caught up now from vacation.

I realize that FLAIR-HUB is far from a simple dataset, but your current flair.py is more than double the length of the previously longest dataset (satlas.py). I think this dataset is pretty comparable to SatlasPretrain in terms of complexity and scope? If we could somehow cut the number of SLOCs in half it would make this much easier to maintain.

gatienc committed:
fixed license name and sorting

gatienc committed:
specific dataset length check

gatienc commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1384 @@
+                labels.
+            domain_years: Optional dictionary mapping domain names to lists of years.
+                If provided, restricts the dataset to only these domain-year combinations.
+                Defaults to None, which uses all available domain-year combinations.
```
Was convenient to scope the data downloaded. I guess I can remove it to make the implementation simpler

gatienc commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1384 @@
+        self,
+        root: Path = 'data',
+        split: str = 'train',
+        split_column: str | None = None,
```
It means that I also have to add split_toy so the typecker accepts it as split_column is already declared by the main datamodule

gatienc commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1384 @@
+            Dictionary containing tensors for each modality. Keys are the
+            modality names: ``mask``, ``AERIAL_RGBI``, ``SPOT_RGBI``,
+            ``DEM_ELEV``, ``AERIAL-RLT_PAN``, ``SENTINEL2_TS``,
+            ``SENTINEL2_MSK-SC``, ``SENTINEL1-ASC_TS``, ``SENTINEL1-DESC_TS``.
```
Do you have an idea on how to provide the data in a "torchgeo" way, A full image with all the modality is not possible as the shape of every modality is very different (  variable temporal data, and very variable size). I am very interested to know your opinion on it.

adamjstewart replied:
Hmm, I guess there is no standard way of handling multi-modal datasets if images can't be stacked into a single image. Unfortunately this means it won't be possible to use any existing trainers in TorchGeo. For Kornia, you would want to prefix each modality name with `mask` or `image` to make it clear which augmentations to apply. Perhaps `image_sentinel1_asc`?

gatienc committed:
Implemented PR recommendation, test WIP

gatienc committed:
dataset download simplification

gatienc committed:
removing some obvious comments and some lines in flair

gatienc committed:
WIP new test

gatienc committed:
fixing trainer tests

gatienc commented:
I tried to reduce the number of SLOC as much as I could, I agree that dataset/flair.py is a very long file : 1204 LOC however on those 1204 LOC the dataset code actually starts only from Line 647, everything before is defining GLOBAL values ( color map, bands etc) and documenting the dataset info.

gatienc commented on tests/data/flair/data.py:
```diff
@@ -0,0 +1,269 @@
+    'D012-195X': ['AF-S1-27', 'AU-S1-18'],
+}
+
+MODALITIES = {
```
There is now 10 files for 1 domain to be able to test the plot method with all modalities. The split file split the data in "train","valid" and "test" so there is 2 other domains used in the test trainers. 

A total of of 22 image are generated : 10 for the first domain for potting and 6 for test training.
Because of the dataset download format ( 1 zip per domain and doubled because of flair-hub toy) but In my opinion it is necessary so we don't have to add lot of path manipulation. 

Compared to last iteration, to keep the code simple I had to re-add DEM and AERIAL-RLT_PAN to the dataset for test training so we don't have to add test coverage on their normalization separately.

Let me know if I should implement it differently

gatienc commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1203 @@
+            'split_5',
+            'split_flairchallenge',
+        ]
+        | Literal['split_toy'] = 'split_1',
```
Here I had to add the  | Literal['split_toy'] even if it doesn't make sense for FLAIRHUB dataset itself (necessary for the FLAIRHUBToy to reuse base class). Do you have an idea on how to implement it in a better way?

adamjstewart replied:
Only way to avoid this would be to have a FLAIRHUBBase dataset with 99% of the code and two tiny FLAIRHUB and FLAIRHUBToy datasets that subclass this.

gatienc commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1203 @@
+            modality names prefixed with ``image_`` or ``mask_``: ``mask``,
+            ``image_aerial_rgbi``, ``image_spot_rgbi``, ``image_dem_elev``,
+            ``image_aerial_rlt_pan``, ``image_sentinel2``,
+            ``mask_sentinel2_snowcloud``, ``image_sentinel1_asc``,
```
I called the sentinel 2 snow cloud data as a mask here? Should I use a different name?

adamjstewart replied:
Nah, feels like a mask to me.

gatienc commented:
To do temporal embedding, the date of year are necessary, I should probably add a list of date of year for temporal data as a key in sample.

gatienc committed:
fix: flairhub toy also has 3 level for crop type

adamjstewart commented on docs/api/datasets/toys.csv:
Usually when git says every line changed, it's due to a change in line ending. Can you double check this?

adamjstewart commented on tests/datasets/test_flair.py:
```diff
@@ -0,0 +1,159 @@
+_DOMAIN_YEARS = {'D006': ['2020'], 'D012': ['2019'], 'D032': ['2019']}
+
+
+class TestFLAIRHUB:
```
Nothing "wrong" with these tests, but they don't look anything like our tests for other datasets, meaning they're harder to maintain than they need to be. See the EuroSAT tests for a good example where we parametrize the dataset fixture over all classes and test basic features like getitem/len/plot, download (good and bad), and any other weird custom stuff.

adamjstewart commented on tests/trainers/test_segmentation.py:
```diff
@@ -96,6 +98,11 @@ def test_trainer(
             case 'ftw':
                 pytest.importorskip('pyarrow')
 
+        if name in ('flairhub_multimodal_landcover', 'flairhub_toy_croptype'):
```
Would be nice if this wasn't required, but I'm okay with this if there isn't an easy way to avoid this. Maybe:

```suggestion
        if name.startswith('flairhub'):
```

gatienc replied:
I put it in the case as I guess is the most expensible method

adamjstewart commented on torchgeo/datamodules/flair.py:
```diff
@@ -0,0 +1,214 @@
+                for batch_key in present_keys:
+                    tensor = batch[batch_key]  # [B, C, H, W]
+                    if tensor.shape[-1] != max_resolution:
+                        tensor = TF.interpolate(
```
Kornia/Pillow also support interpolation. We're trying to reduce our dependence on torchvision as it doesn't seem to be well-maintained anymore.

adamjstewart commented on torchgeo/datamodules/flair.py:
```diff
@@ -0,0 +1,214 @@
+            'split_5',
+            'split_flairchallenge',
+            'split_toy',
+        ] = 'split_1',
```
Remove this, it can be a **kwarg via `split_column`

adamjstewart commented on torchgeo/datamodules/flair.py:
```diff
@@ -0,0 +1,214 @@
+            ),
+        }
+
+    def setup(self, stage: str) -> None:
```
Once you use `**kwargs` for `split_column`, you can delete this method and use the base class implementation

adamjstewart commented on torchgeo/datamodules/flair.py:
```diff
@@ -0,0 +1,214 @@
+            official_splits: Official splits from the research paper.
+                Choose from 'split_1', 'split_2', 'split_3', 'split_4', 'split_5',
+                or 'split_flairchallenge'.
+            concatenate_modalities: If True, concatenate all mono-temporal modalities
+                (AERIAL_RGBI, SPOT_RGBI, DEM_ELEV, AERIAL-RLT_PAN) into an 'image' key.
+                Modalities will be resized to the maximum resolution before concatenation.
+                If False, modalities remain separate. Defaults to False.
```
Neat workaround, but it kinda doubles the complexity to support both. We wouldn't need to override `on_after_batch_transfer` at all if we just always resize/concatenate. Could even do this directly in the dataset and call it a day. I'm not sure who is using FLAIR-HUB in its native resolution in a multi-modal way, it would require a custom Frankenstein model architecture. Can we get the FLAIR folks to comment on this PR and see what they think?

agarioud replied:
Even for modotemporal modalities, we usually use separate backbones to avoid mixing information coming from distinct acquisition dates and having different spectral bandwidths before fusing the representations.

Nonetheless, you are right about the risk of a Frankenstein architecture, and if simplicity is key, since all patches cover the same geographical location, interpolation is possible. No strong opinion on this.

adamjstewart commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1212 @@
+# Copyright (c) TorchGeo Contributors. All rights reserved.
+# Licensed under the MIT License.
+
+"""FLAIRHUB dataset."""
```
```suggestion
"""FLAIR (French Land cover from Aerospace ImageRy) datasets."""
```
We might add #1 and #2 someday, not just HUB.

adamjstewart commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1212 @@
+
+    The dataset is organized by domains (geographical areas) and years.
+    Each domain-year combination has multiple modalities available for
+    download.
+    The full dataset contains
+    66 unique domains (D004-D091, non-consecutive), with years ranging from 2017-2022.
+    Most domains have data for a single year, but some have multiple years available.
```
Formatting is kinda weird, can we combine both sentences in one paragraph?

adamjstewart commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1212 @@
+
+    Set ``download=True`` to automatically download requested modalities from
+    HuggingFace. Only the modalities you select will be downloaded, saving
+    bandwidth and storage.
```
Not needed, this is true of every dataset in TorchGeo and torchvision, it should be self-explanatory

adamjstewart commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1212 @@
+        ]
+        | Literal['split_toy'] = 'split_1',
+        transforms: Callable[[Sample], Sample] | None = None,
+        download: bool = False,
```
We should probably also add sha256 checksums and `checksum: bool = False`. I know this will make the file longer, but it will also be more secure. Note that I have a separate PR to convert all of our existing HF downloads from MD5 to SHA256. Luckily HF tells you the SHA256 of every file so you don't need to download anything.

adamjstewart commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1212 @@
+        """Load a single image.
+
+        Args:
+            path: path to the image
```
I think you need a space between sections? Sphinx isn't complaining because it isn't rendering docs for hidden functions, but still worth fixing.

adamjstewart commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1212 @@
+        ax.imshow(vv_norm, cmap='gray')
+        ax.set_title(title)
+
+    def plot(self, sample: Sample, suptitle: str | None = None) -> Figure:
```
Feels like flair.py is 50% plotting. That's _okay_, but it might be difficult to maintain someday.

adamjstewart commented on torchgeo/datasets/flair.py:
```diff
@@ -0,0 +1,1212 @@
+        # T * C * H * W -> take last timepoint, VV band (index 0)
+        vv = data[-1, 0].numpy()
+        p2, p98 = np.percentile(vv, (2, 98))
+        vv_norm = np.clip((vv - p2) / (p98 - p2 + 1e-6), 0, 1)
```
quantile_normalization

adamjstewart reviewed (commented):
I think we're getting closer here. The dataset is still quite complex, but I'm also close to giving up trying to make it simpler. If we can quickly resolve the remaining change requests before I become swamped with work again, I think we can merge and save additional improvements for future PRs. This dataset is obviously very important, and we would love to have it in TorchGeo, even if it starts out complex and becomes simpler later.

gatienc committed:
cleaner monkeypatching of flairhub domain_years

gatienc committed:
using kornia resize instead of TF

gatienc committed:
Update torchgeo/datasets/flair.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

gatienc committed:
removed official splits to the profit of **kwargs

gatienc committed:
naming of datamodule open to extension

gatienc committed:
docstring fixing

gatienc committed:
quantile normalization for S1

gatienc committed:
fixing end of line for toys.csv

gatienc committed:
adding SHA256 verification option

gatienc committed:
Merge branch 'main' into flairhub

gatienc committed:
fixed missing FLAIRHUb import

adamjstewart commented on tests/trainers/test_segmentation.py:
```diff
@@ -97,6 +99,10 @@ def test_trainer(
         match name:
             case 'ftw':
                 pytest.importorskip('pyarrow')
+            case _ if name.startswith('flairhub'):
```
Seems strange, I would use both names if you want to use case

adamjstewart commented:
Moved this to 0.11 because it seems like there's still a lot of work to do, but if you happen to finish in the next couple days, we could still sneak this into 0.10.

gatienc commented:
I implemented the changes to the tests to be more similar to the EuroSat one, so I think for the Dataset and Test are soon to be done.
The output of the datamodule remains to be discussed @adamjstewart .

adamjstewart commented on tests/tasks/test_segmentation.py:
```diff
@@ -97,6 +99,14 @@ def test_trainer(
+                monkeypatch.setattr(
+                    FLAIRHUB, 'domain_years', {'D006': ['2020'], 'D012': ['2019']}
+                )
+            case 'flairhub_toy_croptype':
```
```suggestion
            case 'flairhub_multimodal_landcover' | 'flairhub_toy_croptype':
```

nilsleh commented on torchgeo/datamodules/flair.py:
```diff
@@ -0,0 +1,171 @@
+from ..datasets.utils import Sample
+from .geo import NonGeoDataModule
+
+# From the research paper : https://arxiv.org/abs/2506.07080 p16
```
I know the differences are quite minor in this case but looking at the referenced table, it seemes that the values here are from the "ALL" partition, when instead normalization statistics should come from the train split only to prevent leakage of information from test set into training process. So I would recommend using the "TRAIN" partition values from the first column in that table

agarioud replied:
Indeed ALL should be discarded, but benchmark metrics from the different papers are using TRAIN+VAL statistics.

isaaccorley replied:
Something to follow up in another PR I think if people feel strongly about it

nilsleh commented:
Otherwise this also LGTM, and rely on people testing this in their training procedures.
